### PR TITLE
Add the obstructed? method for piece.rb

### PR DESCRIPTION
**Pull request by rosiehighsmith**
-install FactoryGirl
-Use RSpec to TDD
-create tests for directional obstructions
-method should identify horizontal, vertical, diagonal obstructions
-method should reject invalid moves

### CODE REVIEW

File: app/models/piece.rb [hunk 1]
```diff
@@ -16,4 +16,57 @@ class Piece < ActiveRecord::Base
...
+
+    def vertical
+      # if y_diff < 0
+
+      # end
+      # elsif y_diff > 0
+
+      # end
+    end
+  end
```
Comment (mscoutermarsh) [inline app/models/piece.rb]: Breaking each direction into its own method makes it easier to think about. :)

File: spec/models/piece_spec.rb [hunk 2]
```diff
@@ -1,5 +1,40 @@
...
 RSpec.describe Piece, type: :model do
-  pending "add some examples to (or delete) #{__FILE__}"
+  describe 'obstructed method' do
+    it 'should return nil if invalid move' do
+      piece = Piece.create(x_position: 4, y_position: 4)
+      expect(piece.obstructed?(1, 3)).to eq nil
+    end
+
+    it 'should return position if obstructed on right' do
+      piece = Piece.create(x_position: 0, y_position: 2)
```
Comment (mscoutermarsh) [inline spec/models/piece_spec.rb]: Instead of using `@board`, you can create pieces here in the correct positions for the test.

```Ruby
game = FactoryGirl.create(:game)
piece_1 = FactoryGirl.create(:piece, x_position: 0, y_position: 2, type: 'Rook', game: game)
piece_2 = FactoryGirl.create(:piece, x_position: 0, y_position: 2, type: 'Rook', game: game)
```

Comment (mscoutermarsh) [inline spec/models/piece_spec.rb]: Then in your code, you can do this

```
game.pieces.exists?(x_position: something, y_position: something) # returns true/false
```

http://api.rubyonrails.org/v4.2.7.1/classes/ActiveRecord/FinderMethods.html#method-i-exists-3F

Comment (jaredAnders) [inline spec/models/piece_spec.rb]: @mscoutermarsh awesome. thanks Mike!

File: app/models/piece.rb [hunk 3]
```diff
@@ -16,4 +16,84 @@ class Piece < ActiveRecord::Base
...
+
+    # set common variables
+    @current_x = x_position
+    @current_y = y_position
+
+    @destination_x = destination_x
+    @destination_y = destination_y
+
+    @x_difference = @current_x - destination_x
+    @y_difference = @current_y - destination_y
```
Comment (mscoutermarsh) [inline app/models/piece.rb]: In models I recommend not using instance variables.

Instead, always pass the data you need as params to methods.

**Reasoning**: Code where the `state` of the data is contained in instance variables can become hard to follow. Imagine if one method updates an instance variable, then another method uses that variable for something. The results of your methods suddenly depend on the order it was called. Makes it hard to know where things change.

Better to have methods that take in parameters, because then you always know that "x input results in y output". 

The way this code is now "x input (+ the value of some instance variables), results in y output"

Make sense? :smile:

File: app/models/piece.rb [hunk 4]
```diff
@@ -16,4 +16,65 @@ class Piece < ActiveRecord::Base
...
+      end
+    else
+      ((destination_x + 1)...x_position).each do |x|
+        return true if game.pieces.exists?(x_position: x, y_position: y_position)
+      end
+    end
+    false
+  end
+
+  def valid_move?(x_difference, y_difference)
```
Comment (mscoutermarsh) [inline app/models/piece.rb]: Can you rename this since we'll been using the same method name this week for each individual piece? Will conflict. 😄

File: app/models/piece.rb [hunk 5]
```diff
@@ -16,4 +16,65 @@ class Piece < ActiveRecord::Base
...
+    x_difference = x_position - destination_x
+    y_difference = y_position - destination_y
+    return true if diagonal?(destination_x, destination_y, x_difference, y_difference)
+    return true if vertical?(destination_y, x_difference, y_difference)
+    return true if horizontal?(destination_x, x_difference, y_difference)
+    return false if valid_move?(x_difference, y_difference)
+    nil
+  end
+
+  def diagonal?(destination_x, destination_y, x_difference, y_difference)
```
Comment (mscoutermarsh) [inline app/models/piece.rb]: All these methods below `obstructed?` can be private

File: app/models/piece.rb [hunk 6]
```diff
@@ -16,4 +16,65 @@ class Piece < ActiveRecord::Base
   def types
     %w(Pawn Rook Knight Bishop Queen King)
   end
+
+  def obstructed?(destination_x, destination_y)
+    x_difference = x_position - destination_x
+    y_difference = y_position - destination_y
+    return true if diagonal?(destination_x, destination_y, x_difference, y_difference)
```
Comment (mscoutermarsh) [inline app/models/piece.rb]: For these method names, I'd do something like `diagonal_obstruction?`

Otherwise, if someone just reads `diagonal`, they might not be sure what it's for

File: app/models/piece.rb [hunk 7]
```diff
@@ -16,4 +16,65 @@ class Piece < ActiveRecord::Base
...
+    return true if vertical?(destination_y, x_difference, y_difference)
+    return true if horizontal?(destination_x, x_difference, y_difference)
+    return false if valid_move?(x_difference, y_difference)
+    nil
```
Comment (mscoutermarsh) [inline app/models/piece.rb]: Should return false at the end aka "This is not a valid move"

File: app/models/piece.rb [hunk 8]
```diff
@@ -16,4 +16,65 @@ class Piece < ActiveRecord::Base
...
+        return true if game.pieces.exists?(x_position: x, y_position: y)
+      end
+    else
+      ((destination_x + 1)...x_position).each.with_index do |x, i|
+        y = destination_y + ((i + 1) * y_direction)
+        return true if game.pieces.exists?(x_position: x, y_position: y)
+      end
+    end
+    false
+  end
```
Comment (mscoutermarsh) [inline app/models/piece.rb]: nicely done 😄

File: db/migrate/20161216223845_add_position_indices_to_pieces.rb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+class AddPositionIndicesToPieces < ActiveRecord::Migration
+  def change
+    add_index :pieces, [:y_position, :x_position], unique: true
```
Comment (mscoutermarsh) [inline db/migrate/20161216223845_add_position_indices_to_pieces.rb]: For this, you can't have a unique constraint. Because this is across the entire table.

So with this constraint, 2 different games wouldn't be able to have a piece in the same position.

If you want to do the unique constraint, this would work

`add_index :pieces, [:game_id, :y_position, :x_position], unique: true`

File: db/migrate/20161217174253_add_game_id_to_pieces.rb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+class AddGameIdToPieces < ActiveRecord::Migration
+  def change
+    add_column :pieces, :game_id, :integer
```
Comment (mscoutermarsh) [inline db/migrate/20161217174253_add_game_id_to_pieces.rb]: add index for this

File: spec/models/piece_spec.rb [hunk 11]
```diff
@@ -1,5 +1,67 @@
 require 'rails_helper'
 
 RSpec.describe Piece, type: :model do
-  pending "add some examples to (or delete) #{__FILE__}"
+  describe 'obstructed method' do
+    it 'should return nil if invalid move' do
+      game = FactoryGirl.create(:game)
+      piece = FactoryGirl.create(:piece, x_position: 4, y_position: 4, type: 'Rook', game: game)
+      expect(piece.obstructed?(1, 3)).to eq nil
```
Comment (mscoutermarsh) [inline spec/models/piece_spec.rb]: `expect(piece.obstructed?(1, 3)).to eq false`

File: spec/models/piece_spec.rb [hunk 12]
```diff
@@ -1,5 +1,67 @@
...
+    end
+
+    it 'should return false if not obstructed on right' do
+      game = FactoryGirl.create(:game)
```
Comment (mscoutermarsh) [inline spec/models/piece_spec.rb]: Since you're using game in every test, we can DRY this up a bit by doing the following.

At the top of the file

`let(:game) { FactoryGirl.create(:game) }`

Then `game` will be available in every test.

more info: http://betterspecs.org/#let

File: spec/models/piece_spec.rb [hunk 13]
```diff
@@ -1,5 +1,67 @@
...
+      piece = FactoryGirl.create(:piece, x_position: 0, y_position: 4, type: 'Rook', game: game)
+      FactoryGirl.create(:piece, x_position: 2, y_position: 2, type: 'Rook', game: game)
+      expect(piece.obstructed?(4, 0)).to eq true
+    end
+
+    it 'should return true if obstructed diagonal up-left' do
+      game = FactoryGirl.create(:game)
+      piece = FactoryGirl.create(:piece, x_position: 4, y_position: 0, type: 'Rook', game: game)
+      FactoryGirl.create(:piece, x_position: 2, y_position: 2, type: 'Rook', game: game)
+      expect(piece.obstructed?(0, 4)).to eq true
```
Comment (mscoutermarsh) [inline spec/models/piece_spec.rb]: 👍 nice job I think you got every case here.

File: spec/models/piece_spec.rb (hunk 11 above)
Comment (mscoutermarsh) [inline spec/models/piece_spec.rb]: Reasoning: we'll let another method handle if the move is on the board or not. :)
